a start for #11 and add rectangles around glyphs

Udi-Fogiel opened the pull request:
This commit is an attempt to add user interface
for configuration of some parameters such as
rule colors, widths and off/on logic per node
type.

It is using the luakeval module, a module which
is format agnostic, so it should work in plain
and LaTeX.

Some parameters are still not implemented.
Mostly options for kern, glue and penalty
should be refined, as they have colors for
different situations, so more keys needs to
be added.

In addition documentation should be updated.

Udi-Fogiel committed:
a start for #11 and add rectangles around glyphs
This commit is an attempt to add user interface
for configuration of some parameters such as
rule colors, widths and off/on logic per node
type.

It is using the luakeval module, a module which
is format agnostic, so it should work in plain
and LaTeX.

Some parameters are still not implemented.
Mostly options for kern, glue and penalty
should be refined, as they have colors for
different situations, so more keys needs to
be added.

In addition documentation should be updated.

Udi-Fogiel commented:
Another thing that should probably added to glyphs is a key to control if the base line should be drawn or not, in fact, boxes around glyphs should probably be drawn above them? so `node.insert_after` in that case?

Udi-Fogiel commented:
oh , and just for testing, the new interface can be used as e.g. `\lvdset{glyph={show=true color={0 0 1 RG}}}`

Udi-Fogiel committed:
add options for kern nodes

pgundlach commented:
```
(/opt/texlive/2022/texmf-dist/tex/generic/iftex/iftex.sty))./lua-visual-debug.l
ua:68: module 'luakeyval' not found:
	no field package.preload['luakeyval']
	[kpse lua searcher] file not found: 'luakeyval'
stack traceback:
	[C]: in function 'require'
	./lua-visual-debug.lua:68: in main chunk
	[C]: in function 'require'
	[\directlua]:1: in main chunk.
\luavisualdebug ... = require("lua-visual-debug")}
                                                  \AtBeginShipout {\directlu...

l.22    \luavisualdebug@dothings
```

My TeX installation is from 2022, so probably too old for luakeyval. Is this standard in newer TeX installations?

Udi-Fogiel committed:
add baseline option for glyphs, draw boxes ontop of glyphs

Udi-Fogiel commented:
> My TeX installation is from 2022, so probably too old for luakeyval. Is this standard in newer TeX installations?

Oh, this is pretty new: https://ctan.org/pkg/luakeyval

You only need this file: https://github.com/Udi-Fogiel/luakeyval/blob/main/luakeyval.lua

Is this dependency a problem?

pgundlach commented:
> Is this dependency a problem?

I am not sure. Does it make sense to include this in lvdebug? Or is this against packaging philosophy?

Udi-Fogiel commented:
Do you mean include the file itself in lvdebug? I don't have objections to it, but maybe CTAN will complain about files with the same name, so probably just rename the file?  but also, why do that? for older TeX distributions?

Or did you mean to ask if using this code is ok in general? It is on CTAN and TeX Live, so up to you?

I'm not sure I follow :)

Udi-Fogiel committed:
use more sane defaults, fix baseline line end y coordinate

Udi-Fogiel committed:
add onlyglyphs option

Udi-Fogiel committed:
recurse into boxes even if they are not visualised

Udi-Fogiel committed:
use pre_shipout_filter in LaTeX to avoid package order problems (#10)

Udi-Fogiel commented:
With the latest commit, the ordering problem reported at #10 is gone. As for the the mysterious line it is more clear with the line `\lvdset{onlyglyphs glyph={show=false} vlist={show=true}}` to only shows vlists. With that line the result is 

<img width="478" height="682" alt="image" src="https://github.com/user-attachments/assets/29c216bf-6698-407a-98ad-43546ef2aff0" />

which reveals that the line is the top of a vbox.

pgundlach commented:
> Do you mean include the file itself in lvdebug? I don't have objections to it, but maybe CTAN will complain about files with the same name, so probably just rename the file? but also, why do that? for older TeX distributions?
> 
> Or did you mean to ask if using this code is ok in general? It is on CTAN and TeX Live, so up to you?
> 
> I'm not sure I follow :)

It was late when I replied...

My concern is the following. Up to now, the Lua file (lua-visual-debug.lua) is sufficient to show the debugging info. Now it requires the file luakeyval. 

If this file is not in the same directory as lua-visual-debug.lua, lua-visual-debug is pretty much useless unless you use TeX-live or something similar. Which can be okay, but if there is a way without this dependency. I'd rather prefer that way. (I use lua visual debug without plain TeX/LaTeX) 

If the key value parsing is done in the style file, and sets some defaults for lua visual debug, that would make me much happier.

I think this should be turned into a module with some defaults, and after requiring the file, these settings could be updated.

Udi-Fogiel committed:
separate key/val interface to a different module

Udi-Fogiel commented:
Ah ok, so in a different lua file that will be loaded from the .sty file? Something like in the last commit? It also make it possible to configure the parameters from the Lua end. I can make things "safer" by adding a metatable for the parameters table, to check user input, but is that really needed?

Udi-Fogiel commented:
Now that I think about it you can also drop the whole key/val parsing from the TeX side and only expose the params table from Lua (like in that latest commit), then users can use that. This way `luakeyval` is not needed.

Udi-Fogiel committed:
fix boxes width and height around (v/h)lists and glyphs

Udi-Fogiel committed:
avoid magic numbers

Udi-Fogiel committed:
use local functions

pgundlach commented:
Looks good to me, now the documentation needs to be updated for the key/value pairs. It might be helpful to put the file  https://github.com/Udi-Fogiel/luakeyval/blob/main/luakeyval.lua into the repository, so people without the package can build the documentation. I won't put it in the CTAN package though.

Udi-Fogiel committed:
add doc for key/val interface

Udi-Fogiel committed:
add opacity key

Udi-Fogiel committed:
add note about opacity

Udi-Fogiel committed:
improve error messages

Udi-Fogiel committed:
add luakeyval.lua

Udi-Fogiel commented:
I've fixed some small issues with the hlist and vlist boxes can you take a look at f677589?

I've added documentation. An example for the usage of the keys would be nice, but I'm not sure what is ok for your workflow.

What about 
```tex
\DocumentMetadata{}
\ExplSyntaxOn
\pdfmanagement_add:nnn { Page / Resources / ExtGState } { lvd } { << /ca~0.5 /CA~0.5 >> }
\ExplSyntaxOff
\documentclass{article}
\usepackage{lua-visual-debug}
\usepackage{unicode-math}
\lvdset{
  glyph = {color={0 0 1 RG} width=0.12} 
  onlyglyphs
  opacity = {/lvd gs}
}
\begin{document}
This is an example $\intop_a^b f(x) dx$
\end{document}
```
<img width="1400" height="197" alt="image" src="https://github.com/user-attachments/assets/09a73032-a98c-4afb-8320-348db18ec2a6" />

Udi-Fogiel commented:
I just noticed: 
https://github.com/pgundlach/lvdebug/blob/c0c2d32a1863f395d4487213b3be6fbb948aa42b/lua-visual-debug.lua#L115
isn't `S` using just the stroke color (just the `RG` part)?

And kern uses `B` which use both but only specify the fill color, is it intentional?

Also, glue and penalty need some more work (I only added to them on/off switch), but it can wait I guess unless you have some idea (I'm not sure what to do there...).

Udi-Fogiel commented:
ok, the Kern part looks intentional, as for rule should I simply update the default value?

Udi-Fogiel committed:
update rule color default value

pgundlach commented:
I think this is good, do you think it should be merged now?

Udi-Fogiel commented:
Yes, I think this is ready. Just note that I did not add an example for the usage of the keys in the doc as I'm not sure what dependencies are ok for the doc, and the changes I did make require the booktabs package (hopefully ok).

Udi-Fogiel commented:
Also feel free to change the keyword names, I don't know if `show` is the best choice, maybe `mark`.

pgundlach commented:
I assume this fixes #11 and #13, is that true (I have to admit that I didn't look into the code close enough, I am busy with other things)

Udi-Fogiel commented:
Yes, should fix #10 as well. I don't believe there is any rush :)

pgundlach commented:
I should have used a squash merge. Sorry for the mess
